zerver/actions/presence.py

```diff
@@ -162,8 +175,65 @@ def do_update_user_presence(
             # last_connected_time >= last_active_time.
             presence.last_connected_time = log_time
             update_fields.append("last_connected_time")
-    if len(update_fields) > 0:
-        presence.save(update_fields=update_fields)
+
+    # WARNING: Delicate, performance-sensitive block.
+
+    # It's time to determine last_update_id and update the presence object in the database.
+    # This briefly takes the crucial lock on the PresenceSequence row for the user's realm.
+    # We're doing this in a single SQL query to avoid any Python-level overhead. We're also
```

Comment:
Avoiding `Python-level overhead` is less the point than avoiding a network round-trip between the application server and the database while holding the lock.

Reply:
Hmm yeah, was thinking of this as part of the "Python-level overhead", but that's not the best phrasing. Updated to
`to avoid any unnecessary overhead, in particular database round-trips,`